pp_hot.c

```diff
@@ -3205,30 +3205,60 @@ PP(pp_match)
+        I32 *l2p = RXp_LOGICAL_TO_PARNO(prog);
+        I32 *p2l_next = RXp_PARNO_TO_LOGICAL_NEXT(prog);
+        SSize_t offs_start, offs_end;
 
         SPAGAIN;			/* EVAL blocks could move the stack. */
-        EXTEND(SP, nparens + i);
-        EXTEND_MORTAL(nparens + i);
-        for (i = !i; i <= nparens; i++) {
-            if (LIKELY(RXp_OFFS_VALID(prog,i)))
-            {
-                const I32 len = RXp_OFFS_END(prog,i) - RXp_OFFS_START(prog,i);
-                const char * const s = RXp_OFFS_START(prog,i) + truebase;
-                if ( UNLIKELY( len < 0 || len > strend - s)
-                )
-                    DIE(aTHX_ "panic: pp_match start/end pointers, i=%ld, "
-                        "start=%ld, end=%ld, s=%p, strend=%p, len=%" UVuf,
-                        (long) i, (long) RXp_OFFS_START(prog,i),
-                        (long)RXp_OFFS_END(prog,i), s, strend, (IV) len);
-                PUSHs(newSVpvn_flags(s, len,
-                    (DO_UTF8(TARG))
-                    ? SVf_UTF8|SVs_TEMP
-                    : SVs_TEMP)
-                );
-            } else {
-                PUSHs(sv_newmortal());
+        EXTEND(SP, logical_nparens + logical_paren);
+        EXTEND_MORTAL(logical_nparens + logical_paren);
+        /* loop over the logical parens in the pattern. This may not
+           correspond to the actual paren checked, as branch reset may
+           mean that there is more than one paren "behind" the logical
+           parens. Eg, in /(?|(a)|(b))/ there are two parens, but one
+           logical paren. */
+        for (logical_paren = !logical_paren;
+             logical_paren <= logical_nparens;
+             logical_paren++)
+        {
+            /* now convert the logical_paren to the physical parens which
+               are "behind" it. If branch reset was not used then
+               physical paren and logical paren are the same */
+            I32 phys_paren = l2p ? l2p[logical_paren] : logical_paren;
+            /* We check the loop invariants below and break out
+             * of the loop explicitly if our checks fail, so use
+             * while (1) here. */
+            while (1) {
+                /* check end offset first, as the start might be >=0 but
+                 * while the end is -1. */
+                if (((offs_end   = RXp_OFFS_END(prog, phys_paren))   != -1) &&
+                    ((offs_start = RXp_OFFS_START(prog, phys_paren)) != -1))
+                {
+                    const SSize_t len = offs_end - offs_start;
+                    const char * const s = offs_start + truebase;
+                    if ( UNLIKELY( len < 0 || len > strend - s) ) {
+                        DIE(aTHX_ "panic: pp_match start/end pointers, paren=%" IVdf ", "
+                            "start=%" IVdf ", end=%" IVdf ", s=%p, strend=%p, len=%" IVdf,
+                            (IV)phys_paren, (IV)offs_start, (IV)offs_end, s, strend, (IV)len);
```

Comment:
You could avoid the casts here by using I32df for the I32 value (phys_paren), and `%zd` for the SSize_t values.

Reply:
ack. willdo also.